Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1

```diff
@@ -544,11 +544,108 @@ function New-ServiceFabricApplicationAction
     {
         Write-Host (Get-VstsLocString -Key SFSDK_CreateApplicationFailed)
         # print application health status if create did not succeed
-        Trace-ServiceFabricApplicationHealth
+        Trace-ServiceFabricApplicationHealth -ApplicationName $ApplicationName
+        throw
+    }
+}
+
+function Start-ServiceFabricApplicationUpgradeAction
+{
+    Param (
+        [hashtable]
+        $UpgradeParameters
+    )
+
+    $global:operationId = $SF_Operations.StartApplicationUpgrade
+    $startAction = { Start-ServiceFabricApplicationUpgrade @UpgradeParameters }
+    $exceptionRetryEvaluator = {
+        param($ex)
+
+        # If upgrade already started, don't retry
+        $upgradeStatus = Get-ServiceFabricApplicationUpgradeAction -ApplicationName $($UpgradeParameters["ApplicationName"])
+        if ($upgradeStatus -and ($upgradeStatus.UpgradeState -ne "RollingBackCompleted" -and $upgradeStatus.UpgradeState -ne "RollingForwardCompleted"))
```

Comment:
You dont need to put braces for the 2nd expression. However, I m not clear on the logic here

Reply:
it is checking if upgrade is already in progress (status should not be RollingBackCompleted or RollingForwardCompleted). if so, it will not retry starting upgrade.

Reply:
But you are returning false in those cases, wouldn't that stop the retry?

Reply:
yes, that is intended. The action being tried here is Start-Upgrade.. If upgrade has already started even though the call failed due to timeout, we would not like to try further.
Only in case of upgrade not started, we want to retry

Reply:
How do we differentiate between the last upgrade vs the current upgrade

Reply:
if current is first upgrade for application, upgrade status will be null, otherwise it will be RollingForwardCompleted or RollingBackCompleted depending upon whether last upgrade was success or rolled back

Reply:
this is the same logic which we use while waiting for upgrade to finish

Reply:
Is there a chance for us to not even attempt retry if the server didn't even accept our retry request.

Reply:
There is nothing like server will not accept request. It will accept and either fail or proceed with upgrade. In either of these cases we don't want to retry.

Reply:
Make sense